### PR TITLE
Add support for setting ``ravenFactory``, ``release``, ``environment`…

### PR DESCRIPTION
**Pull request by bretthoerner**
…` and ``serverName`` via System Environment or Java System Properties.

Related to #244

We haven't discussed this one, but I'd definitely use it and it seemed like an obvious win to me.

---
This change is [Reviewable](https://reviewable.io/reviews/getsentry/raven-java/246)

### CODE REVIEW

**Comment by bretthoerner** [comment]
Review status: 0 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[raven-log4j2/src/main/java/com/getsentry/raven/log4j2/SentryAppender.java, line 115 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQAxWNfnKmZYV-Mpxo0:-KQAxWNfnKmZYV-Mpxo1:bvg9jmt) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-log4j2/src/main/java/com/getsentry/raven/log4j2/SentryAppender.java#L115)):*
> ```Java
>         setServerName(Lookup.lookup("serverName"));
>         setTags(Lookup.lookup("tags"));
>     }
> ```

I should note why I did this in the constructor (for all Appenders and the JUL Handler).

Since most of them are magically configured by their framework calling setters right *after* the constructor, I wanted to set the values from the environment *first* so that any configuration  found in log4j,properties, logback.xml, etc will "win".

In other words, the order for looking up DSN, release, environment, etc goes:

1. configured in your logging framework
2. configured in JNDI
3. configured in OS environment
4. configured in Java System properties

So... since *most* frameworks depend on calling the setters, I wanted the code to look/feel the same and so I chose to do all of the lookups in 2-4 in the constructor, so that the only thing each framework has to do is decide whether it has found a 1 (above) to overwrite any values with.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by tkaemming** [comment]
Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[raven/README.md, line 75 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNYjTvhnh_DkXiSPy4:-KQNYjTwvg6kBKZdbAZE:b-yj18eb) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/README.md#L75)):*
> ```Markdown
> | `SENTRY_SERVERNAME=server1`                                   | Optional, override the server name (rather than looking it up dynamically) |
> | `SENTRY_RAVENFACTORY=com.getsentry.raven.DefaultRavenFactory` | Optional, select the ravenFactory class                                    |
> | `SENTRY_TAGS=tag1:value1,tag2:value2`                         | Optional, provide tags                                                     |
> ```

This doesn't include `SENTRY_EXTRATAGS`. (Maybe this was intentional? I'm also not going to add this comment to the remainder of the `README` files.)

Maybe it'd make sense to collapse this and the following section into a single table that has both the environment variable name as well as the configuration parameter to ensure they're both documented? (Or do you think this would this be too obtuse?)

We also don't seem to document JNDI anywhere — should we?

---

*[raven-log4j/src/main/java/com/getsentry/raven/log4j/SentryAppender.java, line 97 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQN_1qN5RKTOZmAycaP:-KQN_1qN5RKTOZmAycaQ:b2xkk6j) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-log4j/src/main/java/com/getsentry/raven/log4j/SentryAppender.java#L97)):*
> ```Java
>         setEnvironment(Lookup.lookup("environment"));
>         setServerName(Lookup.lookup("serverName"));
>         setTags(Lookup.lookup("tags"));
> ```

Needs `setExtraTags`. (Not going to note this on the remainder of the constructors, either.)

---

*[raven-log4j2/README.md, line 38 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQN_JbZvPXrW_VzTTfo:-KQN_JbZvPXrW_VzTTfp:b-ozacx3) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-log4j2/README.md#L38)):*
> ```Markdown
> <configuration status="warn" packages="org.apache.logging.log4j.core,com.getsentry.raven.log4j2">
>     <appenders>
>         <Raven name="Sentry">
> ```

Can this be self-closing, e.g. `<Raven name="Sentry" />`? (Note: I'm not familiar enough with the parser here to know for sure this works.)

---

*[raven-log4j2/src/main/java/com/getsentry/raven/log4j2/SentryAppender.java, line 115 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQAxWNfnKmZYV-Mpxo0:-KQNb-e0LoYWPlWq7muI:b22mf8j) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-log4j2/src/main/java/com/getsentry/raven/log4j2/SentryAppender.java#L115)):*
<details><summary><i>Previously, bretthoerner (Brett Hoerner) wrote…</i></summary>
> I should note why I did this in the constructor (for all Appenders and the JUL Handler).
> 
> Since most of them are magically configured by their framework calling setters right *after* the constructor, I wanted to set the values from the environment *first* so that any configuration  found in log4j,properties, logback.xml, etc will "win".
> 
> In other words, the order for looking up DSN, release, environment, etc goes:
> 
> 1. configured in your logging framework
> 2. configured in JNDI
> 3. configured in OS environment
> 4. configured in Java System properties
> 
> So... since *most* frameworks depend on calling the setters, I wanted the code to look/feel the same and so I chose to do all of the lookups in 2-4 in the constructor, so that the only thing each framework has to do is decide whether it has found a 1 (above) to overwrite any values with.
</details>
This makes sense. I was going to suggest moving these `Lookup.lookup` calls to the class definition (instead of a parameterless constructor), but then remembered that there's a not insignificant amount of parsing that happens with some of these values (e.g. `tags`) so it'd probably be too complicated. :disappointed:

(… though I guess most of that extraction logic is already encapsulated in static methods?)

---

*[raven-logback/README.md, line 34 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNbwrud-XrPJ91If0I:-KQNbwrud-XrPJ91If0J:b2v7acc) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-logback/README.md#L34)):*
> ```Markdown
> ```xml
> <configuration>
>     <appender name="Sentry" class="com.getsentry.raven.logback.SentryAppender">
> ```

Same question here about self-closing.

---

*[raven-logback/src/main/java/com/getsentry/raven/logback/SentryAppender.java, line 166 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNc49CqFr00gVi0HRp:-KQNc49CqFr00gVi0HRq:b4c9lyr) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-logback/src/main/java/com/getsentry/raven/logback/SentryAppender.java#L166)):*
> ```Java
>         if (RavenEnvironment.isManagingThread()) {
>             return;
>         }
> ```

:thumbsup: 

---

*[raven-logback/src/test/java/com/getsentry/raven/logback/SentryAppenderIT.java, line 20 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNcEfaWdAjkg_6-yhp:-KQNcEfaWdAjkg_6-yhq:b-jg25ju) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-logback/src/test/java/com/getsentry/raven/logback/SentryAppenderIT.java#L20)):*
> ```Java
>     public void setUp() throws Exception {
>         sentryStub = new SentryStub();
>         sentryStub.removeEvents();
> ```

What's the reason for this change?

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 23 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNWlvuXVqd_EMkj0U5:-KQNWlvuXVqd_EMkj0U6:b7dwfg) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L23)):*
> ```Java
>      * @return value of configuration key, if found, otherwise null
>      */
>     public static String lookup(String key) {
> ```

A couple notes/questions about this:

1. To me, it seems like it'd make sense to prefer the Java System Property values over OS environment variables due to specificity, even if that breaks backwards compatibility.
2. This is a stylistic thing, but would it make sense to use `sentry.property.key.foo.bar` style for the property names? (From what I've seen — in my admittedly limited amount of exposure — that seems more conventional?)

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 32 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNc_DlhJQ3p78QCy7A:-KQNc_DmhWuUT254stqk:bgshcug) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L32)):*
> ```Java
>             value = JndiLookup.jndiLookup(key);
>         } catch (ClassNotFoundException | NoClassDefFoundError e) {
>             logger.debug("JNDI not available");
> ```

Might want to pass the `e` value here in case this catches a bug on our side, or restructure this block so that the the class loader check is separate from the lookup call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by bretthoerner** [comment]
Review status: 6 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[raven/README.md, line 75 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNYjTvhnh_DkXiSPy4:-KQfwF_QABiABtIqKvyp:bo10eqy) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/README.md#L75)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> This doesn't include `SENTRY_EXTRATAGS`. (Maybe this was intentional? I'm also not going to add this comment to the remainder of the `README` files.)
> 
> Maybe it'd make sense to collapse this and the following section into a single table that has both the environment variable name as well as the configuration parameter to ensure they're both documented? (Or do you think this would this be too obtuse?)
> 
> We also don't seem to document JNDI anywhere — should we?
</details>
Added extraTags.

I'm not sure about collapsing, I think it's kind of nice to actually show the logging configuration examples (even though I think people should prefer the environment vars when we support them). If anything, maybe we need a new column to also show the Java System Property option (if we're going to change the style)?

As for JNDI... I'll be honest, I've never used it and never heard of anyone who has. It must be popular in enterprise settings? Looks like it was added here: https://github.com/getsentry/raven-java/commit/6afee246

I wonder who actually uses it...

---

*[raven-log4j/src/main/java/com/getsentry/raven/log4j/SentryAppender.java, line 97 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQN_1qN5RKTOZmAycaP:-KQfwDLnO8iyLxHysioM:b-896fix) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-log4j/src/main/java/com/getsentry/raven/log4j/SentryAppender.java#L97)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> Needs `setExtraTags`. (Not going to note this on the remainder of the constructors, either.)
</details>
Done.

---

*[raven-logback/src/test/java/com/getsentry/raven/logback/SentryAppenderIT.java, line 20 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNcEfaWdAjkg_6-yhp:-KQfoeb4KKXTHE2e06Wh:b-nb21g6) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven-logback/src/test/java/com/getsentry/raven/logback/SentryAppenderIT.java#L20)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> What's the reason for this change?
</details>
Creating the appender now does some logging at debug level because (for example) JNDI isn't configured. This means events are waiting in the stub buffer when we test that the length is 0.

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 23 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNWlvuXVqd_EMkj0U5:-KQg2RW7MTBo5C18Yi6Y:b-hent83) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L23)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> A couple notes/questions about this:
> 
> 1. To me, it seems like it'd make sense to prefer the Java System Property values over OS environment variables due to specificity, even if that breaks backwards compatibility.
> 2. This is a stylistic thing, but would it make sense to use `sentry.property.key.foo.bar` style for the property names? (From what I've seen — in my admittedly limited amount of exposure — that seems more conventional?)
</details>
Agreed with both, although I went for `sentry.servername` and not `sentry.server.name` (as an example). It follows a simpler pattern (the same as env vars do, `SENTRY_SERVERNAME`). Thoughts?

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 32 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNc_DlhJQ3p78QCy7A:-KQfxaOKzV0jQisTcGBC:b-896fix) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L32)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> Might want to pass the `e` value here in case this catches a bug on our side, or restructure this block so that the the class loader check is separate from the lookup call.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by bretthoerner** [comment]
Review status: 5 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[raven/README.md, line 75 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNYjTvhnh_DkXiSPy4:-KQg3Y8N6J0GQOwdYflZ:bj7ib6i) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/README.md#L75)):*
<details><summary><i>Previously, bretthoerner (Brett Hoerner) wrote…</i></summary>
> Added extraTags.
> 
> I'm not sure about collapsing, I think it's kind of nice to actually show the logging configuration examples (even though I think people should prefer the environment vars when we support them). If anything, maybe we need a new column to also show the Java System Property option (if we're going to change the style)?
> 
> As for JNDI... I'll be honest, I've never used it and never heard of anyone who has. It must be popular in enterprise settings? Looks like it was added here: https://github.com/getsentry/raven-java/commit/6afee246
> 
> I wonder who actually uses it...
</details>
OK, I think I've handled all feedback except deciding how to handle documentation.

Open issues,

1. JNDI: document or leave it as is for pro-users (if any)?
2. We have 3 (other) ways to config with different keys: `serverName` (log config), `SENTRY_SERVERNAME` (env), `sentry.servername` (properties). Need to decide how we want to document them all. Maybe you're right about a single table, but it's still a lot of info to squeeze in cleanly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by tkaemming** [comment]
Reviewed 11 of 11 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[raven/README.md, line 75 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNYjTvhnh_DkXiSPy4:-KQgA_Qp9I8OJl52neS1:b-g5rbxi) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/README.md#L75)):*
<details><summary><i>Previously, bretthoerner (Brett Hoerner) wrote…</i></summary>
> OK, I think I've handled all feedback except deciding how to handle documentation.
> 
> Open issues,
> 
> 1. JNDI: document or leave it as is for pro-users (if any)?
> 2. We have 3 (other) ways to config with different keys: `serverName` (log config), `SENTRY_SERVERNAME` (env), `sentry.servername` (properties). Need to decide how we want to document them all. Maybe you're right about a single table, but it's still a lot of info to squeeze in cleanly.
</details>
I think it'd be a reasonable compromise to have both the configuration example (sort of as a quick start for the lazy) as well as the table (reference documentation.)

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 23 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNWlvuXVqd_EMkj0U5:-KQgBJSpNfgQBGoNrBwp:b-h69u49) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L23)):*
<details><summary><i>Previously, bretthoerner (Brett Hoerner) wrote…</i></summary>
> Agreed with both, although I went for `sentry.servername` and not `sentry.server.name` (as an example). It follows a simpler pattern (the same as env vars do, `SENTRY_SERVERNAME`). Thoughts?
</details>
Seems reasonable to me.

---

*[raven/src/main/java/com/getsentry/raven/config/Lookup.java, line 36 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgBTH84XePzOrF4uel:-KQgBTH84XePzOrF4uem:bgrmq2n) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/src/main/java/com/getsentry/raven/config/Lookup.java#L36)):*
> ```Java
> 
>         // Try to obtain from a Java System Property
>         String propName = "sentry." + key.toLowerCase();
> ```

Just a stylistic thing, but not sure it's worth assigning a local for the name here and for the environment variable any more.

---

*[raven/src/main/java/com/getsentry/raven/jul/SentryHandler.java, line 103 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgBawU47kWglO-5CgK:-KQgBawV30emKjpn6-3H:b-6omijw) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/src/main/java/com/getsentry/raven/jul/SentryHandler.java#L103)):*
> ```Java
>         setExtraTags(Lookup.lookup("extraTags"));
> ```

Lotta extra newlines here.

---

*[raven/src/main/java/com/getsentry/raven/util/Util.java, line 60 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgBfczXk7sdAmqjKpj:-KQgBfczXk7sdAmqjKpk:buv5qtl) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/src/main/java/com/getsentry/raven/util/Util.java#L60)):*
> ```Java
>      * @return Set of Strings representing extra tags
>      */
>     public static Set<String> parseExtraTags(String extraTagsString) {
> ```

Good move. :thumbsup:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by bretthoerner** [comment]
Review status: 13 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[raven/src/main/java/com/getsentry/raven/jul/SentryHandler.java, line 103 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgBawU47kWglO-5CgK:-KQgKyLtdw7dnxRl2Wdi:b-1j8rzx) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/src/main/java/com/getsentry/raven/jul/SentryHandler.java#L103)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> Lotta extra newlines here.
</details>
Oops, although I did mean to leave 1 to separate the different things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by bretthoerner** [comment]
Review status: 10 of 16 files reviewed at latest revision, 1 unresolved discussion.

---

*[raven/README.md, line 75 \[r1\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQNYjTvhnh_DkXiSPy4:-KQgU9SiHG2RIS-CaSLh:b-896fix) ([raw file](https://github.com/getsentry/raven-java/blob/c4e1d23974d151a9ee3cc4b98a44972dacd2191f/raven/README.md#L75)):*
<details><summary><i>Previously, tkaemming (ted kaemming) wrote…</i></summary>
> I think it'd be a reasonable compromise to have both the configuration example (sort of as a quick start for the lazy) as well as the table (reference documentation.)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by bretthoerner** [comment]
Review status: 10 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[raven/README.md, line 87 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgUQEtyyxhQos5ZtWR:-KQgUQEuAZ1Np1Qqf1-2:bf86lvo) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/README.md#L87)):*
> ```Markdown
> .level=WARN
> handlers=com.getsentry.raven.jul.SentryHandler
> com.getsentry.raven.jul.SentryHandler.dsn=https://publicKey:secretKey@host:port/1?options
> ```

I only removed this to keep it in sync with the table (where it was too long to be readable). I figured people know better without it there? They'll just paste the DSN they get from Sentry imo...?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by tkaemming** [comment]
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[raven/README.md, line 87 \[r2\]](https://reviewable.io:443/reviews/getsentry/raven-java/246#-KQgUQEtyyxhQos5ZtWR:-KQguOFELAdLG75Lbiqx:b-krqt60) ([raw file](https://github.com/getsentry/raven-java/blob/11fb7d8732367a959d2f62acf81e692ddfa885e2/raven/README.md#L87)):*
<details><summary><i>Previously, bretthoerner (Brett Hoerner) wrote…</i></summary>
> I only removed this to keep it in sync with the table (where it was too long to be readable). I figured people know better without it there? They'll just paste the DSN they get from Sentry imo...?
</details>
We can roll with this for now, if it ends up being an issue that causes support tickets we can change it up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/raven-java/246)*

**Comment by tkaemming** [comment]
Whenever you're ready, let's merge and release as 7.6.1.
